File: crates/spfs/src/cli/_cmd_migrate.py
import os
import argparse

import structlog

import spfs
import spfs.storage.fs.migrations

_logger = structlog.get_logger("cli")


def register(sub_parsers: argparse._SubParsersAction) -> None:

    migrate_cmd = sub_parsers.add_parser("migrate", help=_migrate.__doc__)
    migrate_cmd.add_argument(
        "path",
        metavar="PATH",
        help="the path to the filesystem repository to migrate",
    )
    migrate_cmd.add_argument(
        "--upgrade",
        default=False,
        action="store_true",
        help="replace the old data with the migrated data once complete",
    )
    migrate_cmd.set_defaults(func=_migrate)


def _migrate(args: argparse.Namespace) -> None:
    """Migrate the data from and older repository format to the latest one."""

    repo_root = os.path.abspath(args.path)
    if args.upgrade:
        spfs.storage.fs.migrations.upgrade_repo(repo_root)
        result = repo_root
    else:
        result = spfs.storage.fs.migrations.migrate_repo(repo_root)
    _logger.info("migrated", path=result)

File: crates/spfs/src/cli/cmd_migrate.rs
// Copyright (c) 2021 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use clap::Args;
use spfs::Error;

/// Migrate the data from and older repository format to the latest one
#[derive(Debug, Args)]
pub struct CmdMigrate {
    /// Replace old data with migrated data one complete
    #[clap(long)]
    upgrade: bool,

    /// The path to the filesystem repository to migrate
    path: String,
}

impl CmdMigrate {
    pub async fn run(&mut self, _config: &spfs::Config) -> spfs::Result<i32> {
        let repo_root = std::path::PathBuf::from(&self.path)
            .canonicalize()
            .map_err(|err| Error::InvalidPath((&self.path).into(), err))?;
        let result = if self.upgrade {
            spfs::storage::fs::migrations::upgrade_repo(repo_root).await?
        } else {
            spfs::storage::fs::migrations::migrate_repo(repo_root).await?
        };
        tracing::info!(path = ?result, "migrated");
        Ok(0)
    }
}
